### PR TITLE
PLY mesh support

### LINKED ISSUE #24960: PLY models with unknown vertex properties are loaded incorrectly
### System Information

OpenCV version: 5.x
OS, compiler: all

### Detailed description

Each vertex in a PLY file contains properties which are known (like x, y, z) or not known to a PLY loader.
Unknown properties are ignored at header parsing but should be taken into account at data loading to correctly skip unknown properties and load known ones.
For example (see reproducer) the file `viz/dragon.ply` has vertices with properties `x`, `y`, `z`, `red`, `green`, `blue` and unknown property `alpha`.
The parser fails to properly load these vertices since it does not skip the `alpha` property.
Moreover, the properties are always read in the same order without looking at actual properties order given in the header.

### Steps to reproduce

Take a test `PointCloud.LoadSavePly` from `modules/3d/test/test_pointcloud.cpp` and replace filename from `pointcloudio/orig.ply` to `viz/dragon.ply`:
```cpp
TEST(PointCloud, LoadSavePly)
{
    std::vector<cv::Point3f> points;
    std::vector<cv::Point3f> normals;
    std::vector<cv::Point3_<uchar>> rgb;

    auto folder = cvtest::TS::ptr()->get_data_path();
    std::string new_path = tempfile("new.ply");

    cv::loadPointCloud(folder + "viz/dragon.ply", points, normals, rgb);
    cv::savePointCloud(new_path, points, normals, rgb);

    std::vector<cv::Point3f> points_gold;
    std::vector<cv::Point3f> normals_gold;
    std::vector<cv::Point3_<uchar>> rgb_gold;

    cv::loadPointCloud(new_path, points_gold, normals_gold, rgb_gold);

    EXPECT_EQ(normals_gold, normals);
    EXPECT_EQ(points_gold, points);
    EXPECT_EQ(rgb_gold, rgb);
    std::remove(new_path.c_str());
}
```
There will be an error:
```
[ RUN      ] PointCloud.LoadSavePly
/home/savuor/ocv/opencv/modules/3d/test/test_pointcloud.cpp:121: Failure
Expected equality of these values:
  points_gold
    Which is: { ... }
  points
    Which is: { [-0.021492, 0.150723, 0.0996887], [0.300953, -nan, -0.151648], [0.0285347, 0.221372, -0.618525], [-0.589552, -nan, 0.192013], [0.0213919, 0.135382, -0.745135], [-0.102074, -nan, -0.179403],  ... }
/home/savuor/ocv/opencv/modules/3d/test/test_pointcloud.cpp:122: Failure
Expected equality of these values:
  rgb_gold
    Which is: { ... }
  rgb
    Which is: { ... }
[  FAILED  ] PointCloud.LoadSavePly (176 ms)
```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

### PR DESCRIPTION
**Pull request by savuor**
**Warning:** The PR changes exising API.

Fixes #24960
Connected PR: [#1145@extra](https://github.com/opencv/opencv_extra/pull/1145)

### Changes
* Adds faces loading from and saving to PLY files
* Fixes incorrect PLY loading (see issue)
* Adds per-vertex color loading / saving

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
@savuor There are build warnings on Windows.

File: modules/3d/src/pointcloud/io_ply.cpp [hunk 1]
```diff
@@ -142,42 +310,184 @@ T readNext(std::ifstream &file, DataFormat format)
...
+    int val {0};
+    if (format == DataFormat::ASCII)
+    {
+        file >> val;
+    }
+    else
+    {
+        file.read((char *)&val, sizeof(uchar));
+    }
+    return (uchar)(val & 0xff);
```
Comment (asmorkalov) [inline modules/3d/src/pointcloud/io_ply.cpp:325]: The solution looks suspicious.
```
file.read((char *)&val, sizeof(uchar));
```
Depends on endianess. On big-endian you'll get something 0xFF000... and the result will be always zero.

File: modules/3d/src/pointcloud/io_ply.hpp [hunk 2]
```diff
@@ -18,19 +18,37 @@ enum class DataFormat
...
 class PlyDecoder CV_FINAL : public BasePointCloudDecoder
 {
 public:
     void readData(std::vector<Point3f> &points, std::vector<Point3f> &normals, std::vector<Point3_<uchar>> &rgb, std::vector<std::vector<int32_t>> &indices) CV_OVERRIDE;
 
 protected:
     bool parseHeader(std::ifstream &file);
-    void parseBody(std::ifstream &file, std::vector<Point3f> &points, std::vector<Point3f> &normals, std::vector<Point3_<uchar>> &rgb);
+    void parseBody(std::ifstream &file, std::vector<Point3f> &points, std::vector<Point3f> &normals, std::vector<Point3_<uchar>> &rgb,
+                   std::vector<std::vector<int32_t>> &indices);
```
Comment (asmorkalov) [inline modules/3d/src/pointcloud/io_ply.hpp:43]: why not just int?

**Comment by asmorkalov** [comment]
Windows:
```
 [ RUN      ] PointCloud/PlyTest.LoadSaveMesh/0, where GetParam() = "pointcloudio/orig.ply"
[ERROR:0@230.732] global io_ply.cpp:260 cv::PlyDecoder::parseHeader Vertex property x is not presented in the file
[ WARN:0@230.732] global load_point_cloud.cpp:172 cv::saveMesh Have no vertices to save
[ERROR:0@230.732] global io_ply.cpp:36 cv::PlyDecoder::parseHeader Provided file is not in PLY format
C:\GHA-OCV-2\_work\opencv\opencv\opencv\modules\3d\test\test_pointcloud.cpp(177): error: Value of: !indices.empty()
  Actual: false
Expected: true
[  FAILED  ] PointCloud/PlyTest.LoadSaveMesh/0, where GetParam() = "pointcloudio/orig.ply" (1 ms)
```

File: modules/3d/src/pointcloud/io_ply.cpp (hunk 1 above)
Comment (savuor) [reply modules/3d/src/pointcloud/io_ply.cpp:325]: Sure, gonna fix it

File: modules/3d/src/pointcloud/io_ply.hpp (hunk 2 above)
Comment (savuor) [reply modules/3d/src/pointcloud/io_ply.hpp:43]: For consistency purposes
This data type is used everywhere across point cloud / mesh loading for indices

**Comment by asmorkalov** [comment]
The mentioned CI failure on Windows is not a glitch. I reproduced the same behaviour on Windows locally.

File: modules/3d/include/opencv2/3d.hpp [hunk 3]
```diff
@@ -2800,10 +2833,35 @@ CV_EXPORTS_W void loadMesh(const String &filename, OutputArray vertices, OutputA
...
+*  @brief Saves a mesh to a specified file.
+*
+* The function saves mesh to the specified file.
+* File format is chosen based on the filename extension.
+*
+* @param filename Name of the file.
+* @param vertices (vector of Point3f) vertex coordinates of a mesh
+* @param indices (vector of vectors of int) vertex normals of a mesh
+*/
+CV_EXPORTS_W void saveMesh(const String &filename, InputArray vertices, InputArrayOfArrays indices);
```
Comment (asmorkalov) [inline modules/3d/include/opencv2/3d.hpp:2864]: Why do we need overloads. Looks like the most complete function with default values is enough, e.g.:
```
CV_EXPORTS_W void saveMesh(const String &filename, InputArray vertices,  InputArrayOfArrays indices, InputArray normals = noArray(), colors = noArray());
```

**Review by asmorkalov — APPROVED** [review]

File: modules/3d/include/opencv2/3d.hpp (hunk 3 above)
Comment (savuor) [reply modules/3d/include/opencv2/3d.hpp:2864]: Didn't want to change the existing API;
But you're right, this one looks better, fixing it

**Pull request merged**